### PR TITLE
.github: Add dependabot config and thus enable dependabot

### PR DESCRIPTION
**Pull request by mxinden**
With this pull request I suggest to enable [Dependabot](dependabot.com/) for this repository. Dependabot will scan the dependencies used in Substrate (both `Cargo.toml` and `Cargo.lock`) and create pull requests whenever newer versions of those dependencies are available.

We are using Dependabot in [rust-libp2p since about 2 months now](https://github.com/libp2p/rust-libp2p/pull/1744). While annoying at first, I think it is well worth it, reducing the long-term toil keeping dependencies up-to-date as well as staying in sync with recent (security) bug fixes.

Please don't just merge here but signal your consent / dissent for now. Keeping dependencies up-to-date is a lot of work and thus a team effort.

### CODE REVIEW

**Comment by bkchr** [comment]
I'm fine with this, however this pr should already include the changes that updates all dependencies to the newest version. This can be done using cargo outdated. If we don't do this, it will be the hell the next days.

How does it look like when there is a major version and requires a manual update? Will it still notify us?

**Comment by mxinden** [comment]
> I'm fine with this, however this pr should already include the changes that updates all dependencies to the newest version.

Running `cargo update` updates 380 crates, likely most of them transitive dependencies. I would prefer not to update all of these at once because (a) I don't have enough expertise across the codebase to decide on each individual case including doing the necessary manual code changes and (b) tracking down regressions on one huge update, singling out the specific cause, is difficult.

> If we don't do this, it will be the hell the next days.

Dependabot will not create a pull request for each outdated dependency at once, but throttle to at most 5 opened at the same time. See _Features_ section on https://dependabot.com/ for details.

@bkchr are you still in favor of one huge update or are you fine doing this incrementally?

> How does it look like when there is a major version and requires a manual update? Will it still notify us?

Yes, Dependabot will still create a pull request. One can then push to the branch Dependabot creates to do the required updates manually. See https://github.com/libp2p/rust-libp2p/pull/1796 as an example.

File: .github/dependabot.yml
Comment (tomaka) [inline .github/dependabot.yml:6]:
```suggestion
      interval: "daily"
      open-pull-requests-limit: 4
```

I suggest *some* limit, otherwise dependabot will open 380 PRs and we might get yelled at.

File: .github/dependabot.yml
Comment (tomaka) [inline .github/dependabot.yml:4]:
```suggestion
    directory: "/"
    labels: ["A2-insubstantial", "B0-silent", "C1-low"]
```

Also, if we don't add a `labels` entry, it will add a `rust` label, which we probably don't want.
It will automatically add a `rust` label

**Review by tomaka — CHANGES_REQUESTED** [review]

**Comment by tomaka** [comment]
In parallel of adding dependabot, I'd be in favour of doing a single giant `cargo update` in Polkadot, that gets burned in.

File: .github/dependabot.yml
Comment (ordian) [reply .github/dependabot.yml:6]: :+1: 
should be indented at `schedule` level though (2 spaces less)
* https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/configuration-options-for-dependency-updates#open-pull-requests-limit
* https://dependabot.com/docs/config-file/validator/

Comment (mxinden) [reply .github/dependabot.yml:6]: > By default, Dependabot opens a maximum of five pull requests for version updates. Once there are five open pull requests, new requests are blocked until you merge or close some of the open requests. Use open-pull-requests-limit to change this limit. This also provides a simple way to temporarily disable version updates for a package manager.

https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/configuration-options-for-dependency-updates#open-pull-requests-limit

The default value is 5. Is there a particular reason you would prefer 4?

Comment (tomaka) [reply .github/dependabot.yml:7]: I guess no

**Comment by bkchr** [comment]
> In parallel of adding dependabot, I'd be in favour of doing a single giant `cargo update` in Polkadot, that gets burned in.

Isn't the bot really doing an upgrade? So like update everything to the latest version? But that sounds at least like a path to continue.

**Comment by tomaka** [comment]
The bot will update the `Cargo.toml` and `Cargo.lock`, each dependency one by one.

I suggest, in parallel, to do a big `cargo update` in Polkadot. The bot will in the end achieve the same, but it's going to take some time, and the situation of Polkadot is that we have *a lot* of outdated dependencies. The panic we fixed in the 0.26.1 patch release was already fixed back in June and just hasn't been cargo-updated since then.

**Comment by bkchr** [comment]
Yeah I understood this. What I meant is that we should probably do more than a `cargo update`. More a `cargo outdated` run and see what we need to upgrade.

**Review by kirushik — COMMENTED** [review]
I'm OK with this PR in general, but support Pierre in suggesting that we do a huge dependency update round first and only then enable Dependabot.

**Comment by tomaka** [comment]
I believe we can do merge this (after it's been updated to master)

**Review by ordian — APPROVED** [review]

**Comment by mxinden** [comment]
bot merge

**Comment by mxinden** [comment]
For the record, large dependency update via `cargo update` has been merged with https://github.com/paritytech/substrate/pull/7553.

**Comment by mxinden** [comment]
> The PR has approval from two core-devs (or one if the PR is labelled insubstantial).

@bkchr @tomaka would one of you mind reviewing / approving this pull request?

**Review by tomaka — APPROVED** [review]

**Comment by tomaka** [comment]
bot merge

**Pull request merged**
